slam_toolbox/include/slam_toolbox/slam_toolbox_common.hpp

```diff
@@ -120,17 +122,19 @@ class SlamToolbox
   std::unique_ptr<mapper_utils::SMapper> smapper_;
   std::unique_ptr<karto::Dataset> dataset_;
   std::map<std::string, laser_utils::LaserMetadata> lasers_;
+  std::map<std::string, tf2::Transform> m_map_to_odoms_;
 
   // helpers
-  std::unique_ptr<laser_utils::LaserAssistant> laser_assistant_;
-  std::unique_ptr<pose_utils::GetPoseHelper> pose_helper_;
+  std::map<std::string,std::unique_ptr<laser_utils::LaserAssistant>> laser_assistants_;
+  std::vector<std::unique_ptr<pose_utils::GetPoseHelper>> pose_helpers_;
   std::unique_ptr<map_saver::MapSaver> map_saver_;
   std::unique_ptr<loop_closure_assistant::LoopClosureAssistant> closure_assistant_;
   std::unique_ptr<laser_utils::ScanHolder> scan_holder_;
 
   // Internal state
   std::vector<std::unique_ptr<boost::thread> > threads_;
   tf2::Transform map_to_odom_;
+  std::string map_to_odom_child_frame_id_;
```

Comment:
- Previously, map-to-odom TF was always from "/map" to "/odom", so there was no need to keep track of what the child frame was for this TF. Now, the child frame could be "robot1/odom", "robot2/odom", etc., so we need to know what the frame is
- Keeping track of map-to-odom TF, and what the child frame is, is a two step process. Note that these two functions run in separate threads, but are mutex'd, so they will never occur simultaneously:
    - setTransformFromPoses: `map_to_odom_` and `map_to_odom_child_frame_id_` are updated for a specific robot. `map_to_odom_` contains the transform info (xyz, rpy), and `map_to_odom_child_frame_id_` contains the child frame ID (eg. "robot1/odom")
    - publishTransformLoop: Map of child frame to TF, `m_map_to_odoms_`, is updated using `map_to_odom_child_frame_id_` as the key and `map_to_odom_` as the value. This map, then, should provide us with a map to odom TF for all robots. This map is iterated over, and all these TF's are published to the ROS network
- Since setTransformFromPoses updates both `map_to_odom_child_frame_id_` and `map_to_odom_` together, and it's protected by a mutex, the two variables will never go out of sync. It is, however, possible that setTransformFromPoses runs twice (once for robot1, then for robot2), so by the time publishTransformLoop runs, only the second robot (robot2) is added to the map (`m_map_to_odoms_`). 
    - I don't think this is very likely to occur in practice, since publishTransformLoop runs at a much higher frequency than setTransformFromPoses. By default, publishTransformLoop runs 50 times per second, and setTransformFromPoses is throttled by minimum_time_interval, so it runs at a very low frequency. Even if this does occur, the underlying pose graph will always remain correct, the node would just publish a TF for the first robot (robot1) that is out of date since the map failed to update; the next time around, the map would be updated with the most recent TF info from the pose graph, so this error will not accumulate over time
    - That being said, it is still possible and that the first and second robots keep running setTransformFromPoses right next to each other in time. The best way to handle that, then, would be to make another map of child frame IDs to TFs; let's call it `temp_map`. `temp_map` would then be updated with TFs and child frame IDs in setTransformFromPoses, and then publishTransformLoop would use this map to update its own map, `m_map_to_odoms_`
        - If setTransformFromPoses runs only once, then publishTransformLoop behavior is unchanged
        - If setTransformFromPoses runs twice for two different robots, then `temp_map` will have two different TFs for two different robots, so publishTransformLoop would update two elements in `m_map_to_odoms_` and publish both updates
        - If setTransformFromPoses runs twice for the same robot, `temp_map` will only store the latest TF for that robot, so publishTransformLoop would update one element in `m_map_to_odoms_` then just publish that
I'll start working on the map idea in the last couple of bullets.

Reply:
Makes WAY more sense! :smile: